### DISCUSSION TITLE
4.0.0 package and repo structure

### DISCUSSION THREAD

**Discussion by smyrick** [ExpediaGroup/graphql-kotlin#1000] [Ideas] [answered]
I am proposing a new idea for the package structure of the project now that we are expanding more features.

This would be a breaking change since we are moving packages but it could be included in 4.0.0.

I think going forward we should stick with these core principles when creating new package:

* All packages in this repo start with `com.expediagroup.graphql`
* All subpackages should have prefix, ie nothing at root `com.expediagroup.graphql.DeclaredClass`
* When we have generic interfaces and different technology implementations (Spring/Ktor) they should use the same package name in all modules.
  * ie. `client.spring` and `server.spring` 

Given those rules, here is my current layout:

```
graphql-kotlin (com.expediagroup.graphql)
    \ clients
        \ graphql-kotlin-client (.client)
        \ graphql-kotlin-ktor-client (.client.ktor)
        \ graphql-kotlin-spring-client (.client.spring)
    \ generator
        \ graphql-kotlin-schema-generator (.generator)
        \ graphql-kotlin-federation (.generator.federation)
    \ sdl
        \ graphql-kotlin-sdl-generator (.sdl)
        \ graphql-kotlin-hooks-provider (.sdl.hooks)
        \ graphql-kotlin-federated-hooks-provider (.sdl.hooks)
    \ plugins
        \ graphql-kotlin-plugin-core (.plugin)
        \ graphql-kotlin-maven-plugin (.plugin.maven)
        \ graphql-kotlin-gradle-plugin (.plugin.gradle)
    \ servers
        \ graphql-kotlin-server (.server)
        \ graphql-kotlin-spring-server (.server.spring)
```

Any better ideas or better package names?

**Comment by dariuszkuc** [comment, marked as answer]
I think it would make sense to also mention what is the current package as some of the modules already follow this naming pattern ;)

```
graphql-kotlin (com.expediagroup.graphql)
    graphql-kotlin-types (.types) // currently it also exposes com.expediagroup.graphql.server.operations in order to be consistent should it be moved to .types.operations ?
    \ clients
        \ graphql-kotlin-client (.client)
        \ graphql-kotlin-ktor-client (.client.ktor)  // currently it is just .client
        \ graphql-kotlin-spring-client (.client.spring)  // currently it is just .client
    \ generator
        \ graphql-kotlin-schema-generator (.generator) // currently it is just com.expediagroup.graphql
        \ graphql-kotlin-federation (.generator.federation) // currently it is com.expediagroup.graphql.federation
    \ plugins
        \ graphql-kotlin-plugin-core (.plugin)
        \ graphql-kotlin-maven-plugin (.plugin.maven)
        \ graphql-kotlin-gradle-plugin (.plugin.gradle)
    \ servers
        \ graphql-kotlin-server (.server)
        \ graphql-kotlin-spring-server (.server.spring)
```

So from the above it looks like we would need to update the client implementations and generator + federation. I think this change will make packages clearer.

In your above example you moved `sdl-generator` modules to root -> those should only be invoked through plugins so I'd probably keep those under plugins. I'm updating the client generation logic to also use worker API which means updates to `plugin-core` as well. With the above changes I'm thinking plugins directories could look like

```
graphql-kotlin (com.expediagroup.graphql)
    \ plugins
        \ graphql-kotlin-plugin-common* (.plugin)  // rename from core as this would only hold common configuration classes data models
        \ graphql-kotlin-maven-plugin (.plugin.maven)
        \ graphql-kotlin-gradle-plugin (.plugin.gradle)
        \client
            \ graphql-kotlin-client-generator (.plugin.client) // this is current graphql-kotlin-plugin-core module
        \sdl // rename from current schema directory
            \ graphql-kotlin-sdl-generator (.plugin.sdl) // rename from .schema
            \ graphql-kotlin-hooks-provider (.plugin.sdl.hooks)  // rename from .schema.hooks
            \ graphql-kotlin-federated-hooks-provider (.plugin.sdl.hooks) // rename from .schema.hooks

*unsure if we will need the common plugin module, still looking into how to potentially reuse the objects but might end up with separate ones altogether
```

Client and SDL generation logic should only be invoked through plugins so I guess it would make sense to keep those modules under `/plugins`. Alternatively, we could also move `sdl` modules under root (as you proposed) and `client-generator` under `clients`. Unsure how useful that would but, if we do move those (from plugins subdirectory) I think we probably should also expose their functionality through CLI for a complete package.

-- 
Are you also suggesting dropping common root packages (as per https://kotlinlang.org/docs/reference/coding-conventions.html#directory-structure)? I don't think it matters much -> for the most part GH and IDEA will collapse those empty dirs anyway.

**Comment by smyrick** [comment]
* Yes I think we should move the operations in `/types` to either `.types.operations` or `.types.server.operations`

* After talking about the plugins and sdl we have agreed that keep the sdl generator code under the plugins makes the most sense.

* Yes, I am suggesting that we should have all modules be in some directory from the project root to follow that convention. I didn't actually know this was written down but I agree with it.
